src/lib/middleware/demo-authentication.test.ts

```diff
@@ -0,0 +1,57 @@
+import { IAuthType } from '../server-impl';
+import dbInit, { ITestDb } from '../../test/e2e/helpers/database-init';
+import { IUnleashStores } from '../types';
+import { setupAppWithCustomAuth } from '../../test/e2e/helpers/test-helper';
+
+let db: ITestDb;
+let stores: IUnleashStores;
+
+beforeAll(async () => {
+    db = await dbInit('demo_auth_serial');
+    stores = db.stores;
+});
+
+afterAll(async () => {
+    await db?.destroy();
+});
+
+const getApp = (adminLoginEnabled: boolean) =>
+    setupAppWithCustomAuth(stores, () => {}, {
+        authentication: {
+            authDemoAllowAdminLogin: adminLoginEnabled,
+            type: IAuthType.DEMO,
+            createAdminUser: true,
+        },
+    });
+
+test('should create regular user with flag enabled', async () => {
+    const app = await getApp(true);
+    return app.request
+        .post(`/auth/demo/login`)
+        .send({ email: 'test@example.com' })
+        .expect(200)
+        .expect((res) => {
+            expect(res.body.email).toBe('test@example.com');
+            expect(res.body.id).not.toBe(1);
+        });
+});
+
+test('should return 403 for admin user if flag disabled', async () => {
+    const app = await getApp(false);
+    return app.request
+        .post(`/auth/demo/login`)
+        .send({ email: 'admin' })
+        .expect(403);
+});
```

Comment:
As discussed in the other PR thread, I think the right thing to check for here is that using `admin` when the flag is disabled gives you the same response as any other invalid email. I suggest replacing this with a test that compares the admin attempt with a different invalid attempt.

Even if you disagree here, I think 403 is the wrong status code, and that 401 would be better if we were to go with that.

```suggestion
test('if the authDemoAllowAdminLogin flag is disabled, using `admin` should have the same result as any other invalid email', async () => {
    const app = await getApp(false);

    const nonAdminUsername = 'not-an-email'
    const adminUsername = 'admin'

    const nonAdminUser = await app.request
        .post(`/auth/demo/login`)
        .send({ email: nonAdminUsername })

    const adminUser = await app.request
        .post(`/auth/demo/login`)
        .send({ email: adminUsername })

    // verify that the status is the same
    expect(nonAdminUser.status).toBe(adminUser.status);

    for (const user of [nonAdminUser, adminUser]) {
        // verify that the body has the same structure
        expect(user.body).toMatchObject({
            error: expect.stringMatching(/^Could not sign in with /)
        })
    }
});
```